### PR TITLE
refactor: use cache for historical data

### PR DESCRIPTION
**Pull request by EDsCODE**
## Problem

Experimenting with an idea for only refreshing the latest interval

- use cache for any data that is prior to the current interval
- this will enhance query performance for saved items that are recalculated/refreshed without changes to the filters

## Changes

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

File: posthog/queries/trends/trends.py [hunk 1]
```diff
@@ -32,17 +42,52 @@ def _get_sql_for_entity(self, filter: Filter, entity: Entity, team: Team) -> Tup
 
         return sql, params, parse_function
 
-    def _run_query(self, filter: Filter, entity: Entity, team: Team) -> List[Dict[str, Any]]:
-        sql, params, parse_function = self._get_sql_for_entity(filter, entity, team)
+    def get_cached_result(self, filter: Filter, team: Team) -> Optional[List[Dict[str, Any]]]:
+        cache_key = generate_cache_key(f"{filter.toJSON()}_{team.pk}")
+        cached_result_package = get_safe_cache(cache_key)
+        return cached_result_package.get("result") if cached_result_package else None
```
Comment (EDsCODE) [inline posthog/queries/trends/trends.py:48]: Handle if an interval is skipped

File: posthog/queries/trends/trends.py [hunk 2]
```diff
@@ -131,3 +177,28 @@ def _handle_cumulative(self, entity_metrics: List) -> List[Dict[str, Any]]:
         for metrics in entity_metrics:
             metrics.update(data=list(accumulate(metrics["data"])))
         return entity_metrics
+
+
+def is_present_timerange(filter: Filter) -> bool:
```
Comment (EDsCODE) [inline posthog/queries/trends/trends.py:187]: - [ ] test
- [ ] Handle if an interval is skipped

File: ee/clickhouse/views/test/__snapshots__/test_clickhouse_trends.ambr [hunk 3]
```diff
@@ -631,6 +631,78 @@
...
+             (SELECT distinct_id,
+                     argMax(person_id, version) as person_id
+              FROM person_distinct_id2
+              WHERE team_id = 2
+              GROUP BY distinct_id
+              HAVING argMax(is_deleted, version) = 0) AS pdi ON e.distinct_id = pdi.distinct_id
+           WHERE team_id = 2
+             AND event = '$pageview'
+             AND timestamp >= toDateTime('2012-01-14 00:00:00')
+             AND timestamp <= toDateTime('2012-01-15 23:59:59') )
```
Comment (EDsCODE) [inline ee/clickhouse/views/test/__snapshots__/test_clickhouse_trends.ambr:700]: Only reruns on latest interval

File: posthog/queries/trends/trends.py [hunk 4]
```diff
@@ -32,17 +45,70 @@ def _get_sql_for_entity(self, filter: Filter, entity: Entity, team: Team) -> Tup
...
+        return cached_result_package.get("result") if cached_result_package else None
+
+    # Determine if the current timerange is present in the cache
+    def is_present_timerange(self, filter: Filter, team: Team) -> bool:
+        _is_cached = self.get_cached_result(filter, team)
+        if _is_cached and len(_is_cached) > 0:
+            latest_date = _is_cached[0]["days"].pop()
+            parsed_latest_date = parser.parse(latest_date)
+            parsed_latest_date = parsed_latest_date.replace(tzinfo=pytz.timezone(team.timezone))
+            _is_present = is_present_timerange(filter, parsed_latest_date)
```
Comment (neilkakkar) [inline posthog/queries/trends/trends.py:65]: what's happening here? The function signature doesn't seem correct? 🤔

File: ee/clickhouse/views/test/test_clickhouse_trends.py [hunk 5]
```diff
@@ -928,3 +931,138 @@ def test_aggregating_by_group(self):
...
         assert sorted([p["group_key"] for p in curr_people]) == sorted(["org:5", "org:6"])
+
+
+class ClickhouseTestTrendsCaching(ClickhouseTestMixin, LicensedTestMixin, APIBaseTest):
+    maxDiff = None
+    CLASS_DATA_LEVEL_SETUP = False
+
+    @snapshot_clickhouse_queries
+    def test_insight_trends_merging(self):
+        set_instance_setting("CACHE_HISTORY_TRENDS", "all")
```
Comment (neilkakkar) [inline ee/clickhouse/views/test/test_clickhouse_trends.py:942]: this should be `strict_caching_enabled` ?

Comment (neilkakkar) [reply ee/clickhouse/views/test/test_clickhouse_trends.py:942]: nope, `STRICT_CACHING_TEAMS`

Comment (EDsCODE) [reply ee/clickhouse/views/test/test_clickhouse_trends.py:942]: ! why are the snapshots passing if this flag is wrong ...

**Review by neilkakkar — APPROVED** [review]
two issues as mentioned, and I'd optionally love to see some more tests for `merge_results` and `is_present_timerange` (while mocking rest of the parts - so we're confident we're covering everything here), but otherwise LGTM!

I'm guessing at a later stage we can make the interval selection more clever, to handle not just present time range, but arbitrary ones. But this is good enough PoC :)

File: ee/clickhouse/views/test/test_clickhouse_trends.py (hunk 5 above)
Comment (neilkakkar) [reply ee/clickhouse/views/test/test_clickhouse_trends.py:942]: they didn't, they got updated instead 🙈

**Pull request merged**